yetanotherco/aligned_layer
# feat: add onlyAggregator to respondToTask

> [!NOTE]
> ~Merge after #875~
> ~This branch already has 875 contents~, and is running on STAGE

# This PR
`onlyAggregator` modifier was added to `respondToTask`.
For this, `address aggregator` was added on storage (and GAP decreased)

## Implementation

To implement this contract upgrade, the same structure of [implement pausable PR](https://github.com/yetanotherco/aligned_layer/pull/481) was followed;

- A new solidity stored variable was created, `alignedAggregator`.
- This variable is set on a reinitializer(2) function, as the industry standard.
- The variable is also set on the first initialize of the contract, to avoid having to run the reinitialize on a freshly deployed contract (i.e mainnet). This doesn't change the current testnet deploys as this function can not be ran again, thus the reinitializer(2) function.
- 2 new make targets were made, to deploy in devnet and testnet.

## Deploying 

Devnet:
```bash
make anvil_upgrade_add_aggregator
```

Testnet:
- Stage or prod variables should be set accordingly on the `contracts/scripts/.env` file
- Config variables should be set accordingle on: `contracts/script/deploy/config/holesky/aligned.holesky.config.json`
 ```bash
make upgrade_add_aggregator
```

## Testing this PR

1) Deploy previous version of contracts:

 ```bash
git checkout main
```
 ```bash
make anvil_deploy_aligned_contracts
```
 ```bash
make anvil_start_with_block_time
```

The following call should revert:
```bash
cast call 0x1613beB3B2C4f22Ee086B2b38C1476A3cE7f78E8 "alignedAggregator()(address)" --rpc-url 127.0.0.1:8545
```

But the following call should not revert:
(this is just to verify we have the correct ServiceManager address)
```bash
cast call 0x1613beB3B2C4f22Ee086B2b38C1476A3cE7f78E8 "balanceOf(address)(uint256)" 0x1613beB3B2C4f22Ee086B2b38C1476A3cE7f78E8 --rpc-url 127.0.0.1:8545
```

Then close anvil
 ```bash
ctrl + c
```
You must preserve this anvil state before you switch to the new branch, for this you can:
```bash
git stash
```
 ```bash
git checkout 851-add-max-limit-eth-in-payment-from-batcher-to-aggregator-make-the-respond-to-ask-only-callable-by-one-agg
```
 ```bash
git apply
```

2) Upgrade to new version of contracts (make sure you are using previous anvil-state !)

The following upgrade should run and work without problems:
 ```bash
make anvil_upgrade_add_aggregator
```

Then you must start the devnet
 ```bash
make anvil_start_with_block_time
```

And the following call should return the aggregator's address:
```bash
cast call 0x1613beB3B2C4f22Ee086B2b38C1476A3cE7f78E8 "alignedAggregator()(address)" --rpc-url 127.0.0.1:8545
```

### Testing the modifier
To make sure the `onlyAggregator` modifier actually works, you must:
- save a backup of `anvil.aggregator.ecdsa.key.json`
- copy (for example) `operator-3.ecdsa.key.json` with the name `anvil.aggregator.ecdsa.key.json`, where this was before
- run aggregator, operator, batcher.
```bash
make batcher_send_risc0_burst
```
Aggregator should revert when trying to submit:
```bash
Error: execution reverted: custom error 2cbe4195:00000000000000000000000090f79bf6…1aad88f6f4ce6ab8827279cfffb92266 (64 bytes)
```
